### DISCUSSION TITLE
dynamic route takes precedence over static route

### DISCUSSION THREAD

**Discussion by vasanthi-d** [vercel/next.js#37171] [Help]
I am using Next.js 12.

I have created a dynamic route with structure like pages/[category]/[type]/index.js

I also have another route that is static pages/book/[...id].js

Structure looks like this

```
pages
  [category]
    [type]
       index.js --> index page for type
    index.js ---> index page for category
  book
    [...id].js
```

When I type https://localhost:3000/cat1/type1 --> works as expected and displays category index page https://localhost:3000/book/1/2/3 --> works as expected and displays [...id] under book

https://localhost:3000/book --> I expect to display [...id] under book because it is matching static route. Static route should take precedence. But, it is displaying category index page instead.

Can someone please on what am I missing?

**Comment by icyJoseph** [comment]
You need to do `book/[[...id]].js` which is a very edge case, and not very nice looking file name.

From the docs, https://nextjs.org/docs/routing/dynamic-routes#optional-catch-all-routes:

> Catch all routes can be made optional by including the parameter in double brackets ([[...slug]]).
>
> For example, pages/post/[[...slug]].js will match /post, /post/a, /post/a/b, and so on.

**Reply by vasanthi-d** [reply]
@icyJoseph tried with book/[[...id]].js and [https://localhost:3000/book ](https://localhost:3000/book)  displays [[...id]] page under book and treats it as if no parameters are passed.
Is there a way we can display 404 page as there is no 'index' page under 'book' directory itself?

**Reply by icyJoseph** [reply]
If you are using `getStaticProps` or `getServerSideProps`, you could use `ctx.params` to see if there's any and then return `{notFound: true}` as the result of `getStaticProps` or `getServerSideProps`. https://nextjs.org/docs/api-reference/data-fetching/get-static-props#notfound.

If its only client checks you have, I guess you could do `const {asPath} = useRouter()` and then check that `asPath === 'book'`  to show an error page. That'll not return 404 though, the page loads with 200. I guess you somehow check that `id` is a valid list? And return server side 404s?

If the page is optimises at build time, you need to check `asPath` rather than `query`, because query will be empty on first pass.
